Changelog 126 (#88)
intellij-rust/intellij-rust.github.io: _posts/2020-07-12-changelog-126.markdown

---
layout: post
title: "IntelliJ Rust Changelog #126"
date: 2020-07-13 16:54:50 +0300
---

In this release, we enable the new macro expansion engine by default. It was a quite long story for us,

>  so we decided to increase the version of our plugin from `0.2` to `0.3`. Meet the `IntelliJ Rust 0.3` release!

Reviewer: ```suggestion
 so we decided to increase the version of our plugin from `0.2` to `0.3`. Meet the `IntelliJ Rust 0.3`!
```

> In this release, we enable the new macro expansion engine by default. It was a quite long story for us,
>  so we decided to bump the version of our plugin from `0.2` to `0.3`. Meet the `IntelliJ Rust 0.3` release!
>  We'll release a dedicated blog post about the new expansion engine in [CLion blog] soon.

Reviewer: ```suggestion
In this release, we enable the new macro expansion engine by default. It was a quite long story for us,
 so meet the `IntelliJ Rust 0.3`!
 We'll release a dedicated blog post about the new expansion engine in [CLion blog] soon.
```

What do you think? I'm not sure about this change

## New Features

<!-- https://github.com/intellij-rust/intellij-rust/pull/5716 -->
* Enable new macro expansion engine by default. Let's remember why the new engine is better:
  * it supports `impl` blocks and `mod` declarations generated by declarative macros, i.e.
   if some methods are inside `impl` block generated by macro invocation, the plugin makes name resolution

>    and completions for these methods as expected. Ability to make proper name resolution for `mod` declarations
>    provides proper name resolution and completion for crates like `tokio`, `async-std`, `reqwest`, etc.

Reviewer: ```suggestion
   and completions for these methods as expected. Support for generated `mod` declarations
   provides proper name resolution and completion for crates like `tokio`, `async-std`, `reqwest`, etc.
```

  * it's now possible to find macro-generated items by `Go To Struct Declaration` <kbd>Ctrl + N</kbd> or `Go To Symbol`

>    <kbd>Ctrl + Shift + N</kbd>. You can try it with standard atomic types, e.g. `AtomicI16`;

Reviewer: Do we really need this example here?

Author: Hmm, yeah, it's inconsistent a bit

  * it brings semantic highlighting in macro invocations;

>   * most of the IDE features are now works inside macro invocations just like in a regular code: code completion,

Reviewer: ```suggestion
  * most of the IDE features now work inside macro invocations just like in a regular code: code completion,
```

Reviewer: About code completion - is it true? Looks like it's not although it works much better

   `Find Usages` / `Go To Declaration` (<kbd>Ctrl + B</kbd>), `Go To Implementation` (<kbd>Ctrl + Alt + B</kbd>),
   `Go To Type Declaration` (<kbd>Ctrl + Shift + B</kbd>), `Type Info` (<kbd>Ctrl + Shift + P</kbd>),

>    `Extend Selection` (<kbd>Ctrl + W</kbd>), etc

Reviewer: ```suggestion
   `Extend Selection` (<kbd>Ctrl + W</kbd>), etc.
```

>  We'll release a blog post about all these macro-related features soon.

Reviewer: ```suggestion
```

You've already added note about it at the beginning

>  If you ran into an issue with the new engine, you can switch it back to the old one
>  in the plugin settings `Settings | Languages & Frameworks | Rust | Expand Declarative Macros`.

Reviewer: ```suggestion
  If you ran into an issue with the new engine, you can switch it back to the old one
  in the plugin settings `Settings | Languages & Frameworks | Rust | Expand Declarative Macros`.
```

<!-- https://github.com/intellij-rust/intellij-rust/pull/5455 -->

> * Add initial implementation of `Move top-level items` refactoring

Reviewer: ```suggestion
* Add initial implementation of `Move` refactoring (<kbd>F6</kbd>) for top-level items like structs, enums, etc.
```

    {% include gif-img.html path="/assets/posts/changelog-126/move-item" w="700px" %}

<!-- https://github.com/intellij-rust/intellij-rust/pull/5624 -->

> * Initial support of unstable `or_patterns` syntax (see [RFC 2535])

Reviewer: Also, I'd add something like `Note, some code analysis may still not work with such patterns`

* Add an option to sort items alphabetically in `Structure View` (<kbd>Ctrl + F12</kbd>) (by [@t-kameyama])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5652 -->
* Allow extracting non-root modules with `Extract inline module` intention action (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5705 -->
* Update icons for _targets_ in the Cargo tool window
 (see [differences](https://github.com/intellij-rust/intellij-rust/pull/5705))

<!-- https://github.com/intellij-rust/intellij-rust/pull/5693 -->
* Notify users if they use too old Rust toolchain that isn't already supported by the plugin.
 Today we support Rust `1.32.0` or higher.

## Fixes

<!-- https://github.com/intellij-rust/intellij-rust/pull/5612 -->
* Use more precise error annotations for missing/extraneous function call arguments.
 Previously, all function arguments were highlighted (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5653 -->
* Render aliased types for local bindings in the quick documentation <kbd>Ctrl + Q</kbd> (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5689 -->

> * Fix location detection of cargo binaries like `evcxr` (Rust REPL) or `grcov`

Reviewer: ```suggestion
* Fix location detection of cargo binaries like `evcxr` (used by Rust REPL) or `grcov`
```

>  (needed for `Run with Coverage`) on arch linux

Reviewer: ```suggestion
 (needed for `Run with Coverage`)
```

I think it can be reproduced not only on Arch

<!-- https://github.com/intellij-rust/intellij-rust/pull/5683 -->
* Fix `Run external linter` action (was broken in the previous release) (by [@Stzx])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5679 -->
* Fix infinite cargo refresh if you have `openid = 0.4.0` dependency crate

<!-- https://github.com/intellij-rust/intellij-rust/pull/5678 -->

> * Fix false-positive `Lift return` inspection on empty match expr
>  See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5677) for details.

Reviewer: ```suggestion
* Fix false-positive annotation produced by `Lift return` inspection on `match` expression without any arm
```

>  See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5677) for details.

Reviewer: ```suggestion
```

Looks like it's redundant here

<!-- https://github.com/intellij-rust/intellij-rust/pull/5670 -->
* Fix some false-positives in move analysis related to path assignments.
 See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5394) for details.

<!-- https://github.com/intellij-rust/intellij-rust/pull/5669 -->
* Fix some corner cases in doc rendering (<kbd>Ctrl + Q</kbd>).
 See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5631) for details.

<!-- https://github.com/intellij-rust/intellij-rust/pull/5441 -->
* Fix `Run with Coverage`

<!-- https://github.com/intellij-rust/intellij-rust/pull/5701 -->
* Test Tool Window now can be disabled by in registry: <kbd>Ctrl + Shift + A</kbd> -> type "Registry..." ->
 `org.rust.cargo.test.tool.window`.

>  See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5676) for details.

Reviewer: Let's drop it

> * Test Tool Window now can be disabled by in registry: <kbd>Ctrl + Shift + A</kbd> -> type "Registry..." ->
>  `org.rust.cargo.test.tool.window`. (The registry key was removed in the previous release).
>  See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5676) for details.

Reviewer: ```suggestion
* Test Tool Window now can be disabled in registry: <kbd>Ctrl + Shift + A</kbd> -> type "Registry..." ->
 `org.rust.cargo.test.tool.window`.
 See [the issue](https://github.com/intellij-rust/intellij-rust/issues/5676) for details.
```

## Performance Improvements

<!-- https://github.com/intellij-rust/intellij-rust/pull/5674 -->

> * Slightly optimize `RsModulesIndex`: it should slightly speed up highlighting and macro expansion

Reviewer: Do we really want to write about `RsModulesIndex`? Maybe "Slightly speed up highlighting and macro expansion" is enough?

## Internal Improvements

<!-- https://github.com/intellij-rust/intellij-rust/pull/5672 -->
* Add `custom-build` cargo target kind (needed in order to distinguish cargo target for `build.rs`)

<!-- https://github.com/intellij-rust/intellij-rust/pull/5649 -->
<!-- https://github.com/intellij-rust/intellij-rust/pull/5667 -->

> * CI: release toml plugin builds from `GitHub Action` instead of `TeamCity CI`

Reviewer: ```suggestion
* Release Toml plugin builds from `GitHub Action`
```
